### PR TITLE
[SpeechLM2] Add Rotary Time Embedding (RoTE) module

### PR DESCRIPTION
**Pull request by AudranBert**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

This PR adds the support of RoTE defined in https://arxiv.org/abs/2410.12109 and used in [Audio Flamingo Next](https://huggingface.co/nvidia/audio-flamingo-next-hf).
RoTE encodes the absolute timestamp in the features at the output of the encoder.

Question : I have done  x = x.float() in RoTE forward but not sure about this?

**Collection**:  SpeechLM2

# Changelog

- Add Rotary Time Embedding (RoTE) support in SpeechLM2

# Usage

By uncommenting in the examples config, you can train with RoTE:
```yaml
    # rote:
    #   _target_: nemo.collections.speechlm2.modules.rote.RotaryTimeEmbedding
    #   dim: 1024             # must match the encoder output feature dim (must be even)
    #   theta: 1200.0         # base of the geometric frequency bank
    #   rotary_fraction: 0.2  # fraction of dim to rotate
```

or simply :
```python
import torch
from nemo.collections.speechlm2.modules.rote import RotaryTimeEmbedding

dim = 1024
rote = RotaryTimeEmbedding(dim)

x = torch.randn(1, 5, dim)
times = torch.arange(5).float().unsqueeze(0) * 0.08
y = rote(x, times)
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [X] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [X] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [X] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Comment by AudranBert** [comment]
Hi @pzelasko, sorry for the ping :)

File: nemo/collections/speechlm2/modules/rote.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+        Args:
+            x: Feature embeddings of shape ``(B, T, C)`` with ``C == dim`` (channel-last).
+            times: Per-frame absolute time in seconds, shape ``(B, T)`` (broadcastable to ``x[..., 0]``).
+
+        Returns:
+            Tensor of the same shape and dtype as ``x``, rotated by the time-dependent angle.
+        """
+        ori_dtype = x.dtype
+        # OMCAT runs this in fp64, do we need it or fp32 enough?
```
Comment (pzelasko) [inline nemo/collections/speechlm2/modules/rote.py:77]: @AudranBert any thoughts on this? Did you try fp32?

File: examples/speechlm2/conf/salm_automodel.yaml [hunk 2]
```diff
@@ -100,17 +100,26 @@ model:
...
-     modality_adapter:
-       _target_: nemo.collections.speechlm2.modules.perception.IdentityConnector
-       d_model: 1024
-     # spec_augment:
-     #   _target_: nemo.collections.asr.modules.SpectrogramAugmentation
-     #   freq_masks: 2 # set to zero to disable it
-     #   time_masks: 10 # set to zero to disable it
-     #   freq_width: 27
-     #   time_width: 5  # 5 frames = 50ms
+    target: nemo.collections.speechlm2.modules.perception.AudioPerceptionModule
```
Comment (pzelasko) [inline examples/speechlm2/conf/salm_automodel.yaml:103]: It's missing one level of indentation

File: examples/speechlm2/conf/salm.yaml [hunk 3]
```diff
@@ -40,17 +40,26 @@ model:
...
-     modality_adapter:
-       _target_: nemo.collections.speechlm2.modules.perception.IdentityConnector
-       d_model: 1024
-     # spec_augment:
-     #   _target_: nemo.collections.asr.modules.SpectrogramAugmentation
-     #   freq_masks: 2 # set to zero to disable it
-     #   time_masks: 10 # set to zero to disable it
-     #   freq_width: 27
-     #   time_width: 5  # 5 frames = 50ms
+    target: nemo.collections.speechlm2.modules.perception.AudioPerceptionModule
```
Comment (pzelasko) [inline examples/speechlm2/conf/salm.yaml:43]: It's missing one level of indentation

**Review by pzelasko — COMMENTED** [review]
Great contribution overall, left a few comments

**Comment by pzelasko** [comment]
Did you run some experiments on this vs other positional encoding schemes? did you see any improvement in practice?

**Comment by AudranBert** [comment]
> Did you run some experiments on this vs other positional encoding schemes? did you see any improvement in practice?

Hi @pzelasko, thanks for the quick feedback!
Currently, we are running full experiments on our data, and we will have the results next week.
And we did run small experiments with and without RoTE on a specific training set where we thought it will have an impact. 
The datasets are modified version of SLUE-SQA5 where we :
- ask for a word at a given timestamp (time2word)
- ask for the timestamps of a given word (word2time)
- ask for a sentence at a given timestamp (time2sentence)
- ask for the sentence where a given word is (word2sentence)
Here is the plot:
temporal_table

The baseline has not trained on such data and doesn't have RoTE, that's why it is not good. The others have trained on these synthetic datasets and one have RoTE, the other dont.
As you can see RoTE increases the performances by a lot in the word2sentence and word2time tasks.

File: nemo/collections/speechlm2/modules/rote.py (hunk 1 above)
Comment (AudranBert) [reply nemo/collections/speechlm2/modules/rote.py:77]: The experiments we launched ran in FP32; we haven't tried FP64. See below for the results.

File: examples/speechlm2/conf/salm_automodel.yaml (hunk 2 above)
Comment (AudranBert) [reply examples/speechlm2/conf/salm_automodel.yaml:103]: You sure? I just removed one space because it had 3 spaces instead of 2 (all the rest of the file only has 2 spaces as indent)

Comment (pzelasko) [reply examples/speechlm2/conf/salm_automodel.yaml:103]: oh, you're right

**Comment by pzelasko** [comment]
Looks good! Yeah let's merge it. Can you resolve the conflicts?

**Review by pzelasko — DISMISSED** [review]
LGTM

**Comment by pzelasko** [comment]
/ok to test [cb2ca49](https://github.com/NVIDIA-NeMo/NeMo/pull/15794/commits/cb2ca4942d921f987731ea3ce392bebd7f060dfe)

**Comment by AudranBert** [comment]
> Looks good! Yeah let's merge it. Can you resolve the conflicts?

The SpeechLM2 tests are passing after the merge, I saw that there were some changes in encoder_chunking that have been made that did a similar thing (chunk_spans) so used it to compute time_offsets.

**Comment by AudranBert** [comment]
Hi @pzelasko I removed the unused import. It has dismissed your review, i don't know if it is important

**Review by pzelasko — APPROVED** [review]

**Comment by pzelasko** [comment]
/ok to test [f10c335](https://github.com/NVIDIA-NeMo/NeMo/pull/15794/commits/f10c3350a1e6d950866ef2e9fa8b384ea7f35a0c)
